pdi-execution-engine/api/src/main/java/org/pentaho/di/engine/api/reporting/IProgressReporting.java

```diff
@@ -0,0 +1,98 @@
+package org.pentaho.di.engine.api.reporting;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Created by hudak on 1/11/17.
+ */
+public interface IProgressReporting {
+  <S extends IReportingEventSource, D extends Serializable>
+  Publisher<IReportingEvent<S, D>> eventStream( S source, Class<D> type );
+
+  Collection<IReportingEventSource> getReportingSources();
```

Comment:
Yes, could be. Didn't want to assume the implementations had save `hashCode` and `equals` implementations, though we could prescribe that.